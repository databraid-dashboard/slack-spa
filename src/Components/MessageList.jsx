// @flow

/* eslint-disable */
import React, { Component } from 'react';
import { List } from 'semantic-ui-react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { fetchMessagesForChannel } from '../Actions/index';
import Message from './Message';
// import owl from '../images/avatars/owl.png';
import type { Dispatch, OwnProps, State } from '../FlowTypes/';

import injectWidgetId from '../Utils/utils';

export class MessageList extends Component {
  props: {
    selectedChannel: ?string,
    messages: {},
    fetchMessagesForChannel: Function,
  };

  render() {
    let { messages, selectedChannel, fetchMessagesForChannel } = this.props;

    if (!messages) {
      if (selectedChannel) {
        fetchMessagesForChannel(selectedChannel);
      }
      messages = {};
    }

    const messageIds = Object.keys(messages);

    const sizes = ['mini', 'tiny', 'small', 'large', 'big', 'huge', 'massive'];

    /*
NOTE: Properties available for each message:
 - messageId* (integer): Primary key from database table. *Only a property of messages retrieved fmor the database, not on messages emitted via Socket.io
 - avatarImage (string): Image size 24
 - name (string): "Real name", e.g. 'Kurtis Houser'
 - userName (string): "User name", e.g. 'thekurtishouser' or other
 - text (string): The message
 - timestamp (string): Formatted timestamp, e.g. '2017-08-01T22:20:43.643Z'
 - rawTimestamp (string): Raw timestamp e.g. '1501626043.643661'
 - channelName (string): Name of the the channel, e.g. 'dev'
 - statusEmoji (string): As labeled internally by slack, e.g. ':slack:'
*/

    return (
      <List celled size={sizes[2]} className="scrolling">
        {messageIds.map(msgId => {
          const { avatarImage, name, text, timestamp } = messages[msgId];
          return (
            <Message
              key={timestamp}
              avatarImage={avatarImage}
              name={name}
              text={text}
              timestamp={timestamp}
            />
          );
        })}
      </List>
    );
  }
}

export const mapStateToProps = (state: State, ownProps: OwnProps) => {
  let id = ownProps.widgetId;
  const selectedChannel = state.widgets.byId[id].selectedChannel;
  const messages = state.widgets.byId[id].channelData[selectedChannel];
  return {
    selectedChannel,
    messages,
    // currentScore,
  };
};

export const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators({ fetchMessagesForChannel }, dispatch);

export default injectWidgetId(connect(mapStateToProps, mapDispatchToProps)(MessageList));
