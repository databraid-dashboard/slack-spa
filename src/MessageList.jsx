// @flow

import React from 'react';
import { List } from 'semantic-ui-react';
import Message from './Message';
import owl from './images/avatars/owl.png';

import store from './store';
import Actions from './Actions';

export default class MessageList extends React.Component {
  componentWillMount() {
    store.subscribe(() => this.forceUpdate());
  }

  props: {
    selectedChannel: ?string,
  }

  render() {
    const {selectedChannel} = this.props;
    let messages = store.getState().channelData[this.props.selectedChannel];
    if (!messages) {
      if (selectedChannel) {
        setTimeout(
          () => store.dispatch(Actions.fetchMessagesForChannel(selectedChannel)),
          0
        );
      }
      messages = {};
    }

    const messageIds = Object.keys(messages);

    return (
      <List celled>
        {messageIds.map((msgId) => {
          const { avatarImage, name, text, timestamp } = messages[msgId];
          return (
            <Message
              key={msgId}
              avatarImage={owl}
              name={name}
              text={`${avatarImage} says: ${text}`}
              timestamp={timestamp}
            />
          );
        })}
      </List>
    );
  }
}
