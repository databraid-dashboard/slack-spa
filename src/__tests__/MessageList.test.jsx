import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import MessageList from '../Components/MessageList';

describe('Renders message list with default props', () => {
  it('should render a message list when given props', () => {
    const messageList = shallow(<MessageList selectedChannel="dev" />);
    expect(toJson(messageList)).toMatchSnapshot();
  });
});
