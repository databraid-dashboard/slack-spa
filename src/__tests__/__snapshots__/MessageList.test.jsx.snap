// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders message list with default props should render a message list when given props 1`] = `
<List
  celled={true}
  className="scrolling"
  size="small"
/>
`;
